waspc/src/Wasp/Generator/DbGenerator.hs

```diff
@@ -73,6 +88,37 @@ deleteGeneratedMigrationsDirIfRedundant spec projectRootDir = do
   where
     projectMigrationsDirAbsFilePath = SP.fromAbsDir $ projectRootDir </> dbRootDirInProjectRootDir </> dbMigrationsDirInDbRootDir
 
+-- | This function warns the user to run `wasp db migrate-dev` under certain conditions. But first, a few preliminaries:
+--   - This function relies on the latest schema.prisma, and thus should run after the writing of FileDrafts has occurred.
+--   - A non-empty schema.prisma should always exist in the generated project dir, even if the user has no entities defined.
+--     `wasp new` will create one with a `datasource db {}` and `generator client {}`, for example.
+--   - A schema.prisma.wasp-checksum file is created/updated each time `wasp db migrate-dev` is run locally and contains checksum(schema.prisma).
+--
+-- Given that, there are two cases in which we wish to warn the user to run `wasp db migrate-dev`:
+-- (1) If schema.prisma.wasp-checksum exists, but is not equal to checksum(schema.prisma), we know they made changes to schema.prisma and should migrate.
+-- (2) If schema.prisma.wasp-checksum does not exist, but the user has entities defined in schema.prisma (and thus, AppSpec).
+--     This could imply they have never migrated locally, or that they have but are simply missing their generated project dir.
+--     Common scenarios for the second warning include:
+--       - After a fresh checkout, or after `wasp clean`; possible false positives in these cases, but for safety, it's still preferable to warn.
+--       - When they previously had no entities and just added their first.
+checkIfDbSchemaWasUsedForLastMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
+checkIfDbSchemaWasUsedForLastMigration spec projectRootDir = do
+  dbSchemaChecksumFileExists <- doesFileExist dbSchemaChecksumFp
+
+  if dbSchemaChecksumFileExists
+    then do
+      Hex dbSchemaFileChecksum <- checksumFromFilePath dbSchemaFp
+      dbChecksumFileContents <- readFile dbSchemaChecksumFp
+      return $ warnIf (dbSchemaFileChecksum /= dbChecksumFileContents) "Your Prisma schema has changed, you should run `wasp db migrate-dev`."
+    else return $ warnIf entitiesExist "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."
+  where
+    dbSchemaFp = SP.fromAbsFile $ projectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ projectRootDir </> dbSchemaChecksumFileInProjectRootDir
+    entitiesExist = not . null $ getEntities spec
+
+    warnIf :: Bool -> String -> [GeneratorWarning]
+    warnIf b msg = if b then [GenericGeneratorWarning msg] else []
```

Comment:
This was the snippet I was referring to during standup. Mainly there to help cleanup another if/else, but pretty close to defining a variant of `warn` like some other libs have done:

```haskell
when' :: (Monad m, Monoid a) => Bool -> m a -> m a
when' b ma = if b then ma else return mempty
```

But wasn't sure if I wanted to go that far, and start adding a ton to Util.hs so I just kept it outside of IO and local to the user. Better styles/approaches?

Reply:
Oho cool Haskell stuff there!

I have never seen this -> I think what you did is pretty nice and it makes sense to keep it localized for now, and if you realize you require a more general approach, we will probably be smarter regarding how to go about it.

when' is interesting -> although I wonder how it could be used in your situation, due to this combo of Monad and Monoid, I don't think it would be easy to use it directly hm?

Reply:
Cool thanks, I will leave the helper localized to keep it a bit DRYer. In the `when'` case I could replace something like:

```haskell
else return $ warnIf entitiesExist "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."
```

with:

```haskell
else when' entitiesExist (return [GenericGeneratorWarning "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."])
```

But not really worth adding to Util IMO just yet.